llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -295,20 +438,74 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
   return Res;
 }
 
-static string_vector
-saveSpecConstantIDMaps(const std::vector<SpecIDMapTy> &Maps) {
-  string_vector Res;
+// Each fallback device library corresponds to one bit in "require mask" which
+// is an unsigned int32. getDeviceLibBit checks which fallback device library
+// is required for FuncName and returns the corresponding bit. The corresponding
+// mask for each fallback device library is:
+// fallback-cassert:      0x1
+// fallback-cmath:        0x2
+// fallback-cmath-fp64:   0x4
+// fallback-complex:      0x8
+// fallback-complex-fp64: 0x10
+static uint32_t getDeviceLibBits(const std::string &FuncName) {
+  auto DeviceLibFuncIter = DeviceLibFuncMap.find(FuncName);
+  return ((DeviceLibFuncIter == DeviceLibFuncMap.end())
+              ? 0
+              : 0x1 << (DeviceLibFuncIter->second - cl_intel_devicelib_assert));
+}
+
+// For each device image module, we go through all functions which meets
+// 1. The function name has prefix "__devicelib_"
+// 2. The function has SPIR_FUNC calling convention
+// 3. The function is declaration which means it doesn't have function body
+static uint32_t getModuleReqMask(const Module &M) {
+  // Device libraries will be enabled only for spir-v module.
+  if (M.getTargetTriple().substr(0, 7) != "spir64-")
```

Comment:
this can crash if triple is too short. .startswith would work better, but see below.

Reply:
Done.